### PR TITLE
simpler incremental better backedges

### LINKED ISSUE #22125: "invalid age range update" assertion error, present in 0.6RC2 but not 0.6RC1
Apologies again for lack of reduced code snippet since its a tough one to reduce. But I'm seeing the following error (full stack trace [here](https://gist.github.com/marius311/839a310d97879d057e4b79b873ba1b0c))

```
WARNING: An error occurred during inference. Type inference is now partially disabled.
Base.MethodError(f=typeof(Core.Inference.convert)(), args=(Base.AssertionError, "invalid age range update"), world=0x0000000000000ac2)
```

Its reproducible every time on 0.6RC2, but is not present on 0.6RC1.

### LINKED ISSUE #22158: Precompilation error on 0.6
I am experiencing a `signal (6)` during precompilation of the following two modules. This is on `v0.6.0-rc2` built with `make debug`.
```julia
__precompile__()

module A

  import Base: cmp, hash, >

  export hash, typeA
    
  type fmpz end 

  type typeA end 

  >(x::fmpz, y::Int) = cmp(x, y) > 0 

  function hash(a::typeA, h::UInt)
    d = den(a)
    return h
  end 
end
```
```julia
__precompile__()

module B
  using A

  import A: hash

  type typeB 
    y::typeA
  end 

  hash(x::typeB) = hash(x.y)
end
```
```julia
~/julia/julia-git ((v0.6.0-rc2))$ ./usr/bin/julia-debug
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-rc2.0 (2017-05-18 02:31 UTC)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |  x86_64-linux-gnu

julia> using B
INFO: Precompiling module B.
julia-debug: /home/thofmann/julia/julia-git/src/gf.c:1326: jl_method_instance_add_backedge: Assertion `callee->min_world <= caller->min_world && callee->max_world >= caller->max_world' failed.

signal (6): Aborted
while loading no file, in expression starting on line 0
raise at /build/eglibc-oGUzwX/eglibc-2.19/signal/../nptl/sysdeps/unix/sysv/linux/raise.c:56
abort at /build/eglibc-oGUzwX/eglibc-2.19/stdlib/abort.c:89
__assert_fail_base at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:92
__assert_fail at /build/eglibc-oGUzwX/eglibc-2.19/assert/assert.c:101
jl_method_instance_add_backedge at /home/thofmann/julia/julia-git/src/gf.c:1326
jl_insert_backedges at /home/thofmann/julia/julia-git/src/dump.c:2229
_jl_restore_incremental at /home/thofmann/julia/julia-git/src/dump.c:3282
jl_restore_incremental at /home/thofmann/julia/julia-git/src/dump.c:3321
_include_from_serialized at ./loading.jl:157
_require_from_serialized at ./loading.jl:200
_require at ./loading.jl:491
require at ./loading.jl:398
unknown function (ip: 0x7f842be096cb)
jl_call_fptr_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:337
jl_call_method_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:356
jl_apply_generic at /home/thofmann/julia/julia-git/src/gf.c:1930
jl_apply at /home/thofmann/julia/julia-git/src/julia.h:1422
eval_import_path_ at /home/thofmann/julia/julia-git/src/toplevel.c:401
eval_import_path at /home/thofmann/julia/julia-git/src/toplevel.c:428
jl_toplevel_eval_flex at /home/thofmann/julia/julia-git/src/toplevel.c:493
jl_toplevel_eval at /home/thofmann/julia/julia-git/src/toplevel.c:598
jl_toplevel_eval_in at /home/thofmann/julia/julia-git/src/builtins.c:496
eval at ./boot.jl:235
unknown function (ip: 0x7f842bc62b8f)
jl_call_fptr_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:337
jl_call_method_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:356
jl_apply_generic at /home/thofmann/julia/julia-git/src/gf.c:1930
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x7f842bd0d10f)
jl_call_fptr_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:337
jl_call_method_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:356
jl_apply_generic at /home/thofmann/julia/julia-git/src/gf.c:1930
macro expansion at ./REPL.jl:97 [inlined]
#1 at ./event.jl:73
unknown function (ip: 0x7f840f6955af)
jl_call_fptr_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:337
jl_call_method_internal at /home/thofmann/julia/julia-git/src/julia_internal.h:356
jl_apply_generic at /home/thofmann/julia/julia-git/src/gf.c:1930
jl_apply at /home/thofmann/julia/julia-git/src/julia.h:1422
start_task at /home/thofmann/julia/julia-git/src/task.c:267
unknown function (ip: 0xffffffffffffffff)
Allocations: 3155595 (Pool: 3154319; Big: 1276); GC: 4
Aborted (core dumped)
```
This showed up after https://github.com/JuliaLang/julia/commit/5e709e46458f8a1cc566d75718b385a63ac9d917.

### PR DESCRIPTION
**Pull request by vtjnash**
This uses a much simpler method of restoring backedges, so that we don't have the troublesome edges cases exposed by the old method.

fix #22158
fix #22125

### CODE REVIEW

File: src/dump.c [hunk 1]
```diff
@@ -1257,31 +1258,48 @@ static void jl_serialize_lambdas_from_mod(jl_serializer_state *s, jl_module_t *m
...
+// flatten the backedge map reachable from caller into callees
+static void jl_collect_backedges_to(jl_serializer_state *s, jl_method_instance_t *caller, jl_array_t *callees, arraylist_t *to_restore)
+{
+    jl_array_t **pcallee = (jl_array_t**)ptrhash_bp(&edges_map, (void*)caller),
+                *callee = *pcallee;
+    if (callee != HT_NOTFOUND) {
+        arraylist_push(to_restore, (void*)pcallee);
+        arraylist_push(to_restore, (void*)callee);
+        *pcallee = (jl_array_t*) HT_NOTFOUND;
+        jl_array_ptr_1d_append(callees, callee);
```
Comment (JeffBezanson) [inline src/dump.c]: `callee` sounds singular to me, so it's a bit confusing that it's an array. Could use a comment explaining what it's an array of, and/or a variable rename. Is it an array of things called by `caller`?

**Comment by vtjnash** [comment]
(backport branch created at jn/release0.6-incremental-better-backedges)

File: src/dump.c [hunk 2]
```diff
@@ -2175,12 +2175,14 @@ static void jl_insert_methods(jl_array_t *list)
 static size_t lowerbound_dependent_world_set(size_t world, arraylist_t *dependent_worlds)
 {
     size_t i, l = dependent_worlds->len;
+    if (world <= (size_t)dependent_worlds->items[l - 1])
+        return world;
     for (i = 0; i < l; i++) {
```
Comment (yuyichao) [inline src/dump.c]: So this is sorted and can't be empty? Then only check the first element?

Comment (vtjnash) [inline src/dump.c]: The value of the return value is also meaningful. (This is doing std::lower_bound on a sorted vector, not just a range test)

Comment (vtjnash) [inline src/dump.c]: The value of the return value is also meaningful. (This is doing std::lower_bound on a sorted vector, not just a range test)

Comment (vtjnash) [inline src/dump.c]: The value of the return value is also meaningful. (This is doing std::lower_bound on a sorted vector, not just a range test)

Comment (yuyichao) [inline src/dump.c]: OK. I think I got the sorting order wrong. Though in that case this should probably be a bisect?

Comment (yuyichao) [inline src/dump.c]: OK. I think I got the sorting order wrong. Though in that case this should probably be a bisect?

Comment (vtjnash) [inline src/dump.c]: yes, but it wouldn't be faster, so no

**Comment by StefanKarpinski** [comment]
Notes: @tkelman will run PkgEval on this against the release-0.6 branch as soon as it passes CI here. If that looks good, this addresses https://github.com/JuliaLang/julia/issues/22158, which is the last remaining issue blocking 0.6.0 (as of RC3).

**Pull request merged**